node.go

```diff
@@ -0,0 +1,325 @@
+		return false, err
+	} else if !deleted {
+		return false, nil
+	}
+
+	if subn.empty() {
+		n.links[subi] = nil
+	} else {
+		ln.dirty = true
+	}
+
+	return true, nil
+}
+
+func (n *node) forEachAt(ctx context.Context, bs cbor.IpldStore, height int, start, offset uint64, cb func(uint64, *cbg.Deferred) error) error {
+	if height == 0 {
+		for i, v := range n.values {
+			if v != nil {
+				ix := offset + uint64(i)
+				if ix < start {
+					continue
+				}
+
+				if err := cb(offset+uint64(i), v); err != nil {
+					return err
+				}
+			}
+		}
+
+		return nil
+	}
+
+	subCount := nodesForHeight(height)
+	for i, ln := range n.links {
+		if ln == nil {
+			continue
+		}
+		subn, err := ln.load(ctx, bs, height-1)
+		if err != nil {
+			return err
+		}
+
+		offs := offset + (uint64(i) * subCount)
+		nextOffs := offs + subCount
+		if start >= nextOffs {
+			continue
+		}
+
+		if err := subn.forEachAt(ctx, bs, height-1, start, offs, cb); err != nil {
+			return err
+		}
+	}
+	return nil
+
+}
+
+var errNoVals = fmt.Errorf("no values")
+
+func (n *node) firstSetIndex(ctx context.Context, bs cbor.IpldStore, height int) (uint64, error) {
+	if height == 0 {
+		for i, v := range n.values {
+			if v != nil {
+				return uint64(i), nil
+			}
+		}
+		// Empty array.
+		return 0, errNoVals
+	}
+
+	for i, ln := range n.links {
+		if ln == nil {
+			// nothing here.
+			continue
+		}
+		subn, err := ln.load(ctx, bs, height-1)
+		if err != nil {
+			return 0, err
+		}
+		ix, err := subn.firstSetIndex(ctx, bs, height-1)
+		if err != nil {
+			return 0, err
+		}
+
+		subCount := nodesForHeight(height)
+		return ix + (uint64(i) * subCount), nil
+	}
+
+	return 0, errNoVals
+}
+
+func (n *node) set(ctx context.Context, bs cbor.IpldStore, height int, i uint64, val *cbg.Deferred) (bool, error) {
+	if height == 0 {
+		alreadySet := n.values[i] != nil
+		n.values[i] = val
+		return !alreadySet, nil
+	}
+
+	nfh := nodesForHeight(height)
+
+	// Load but don't mark dirty or actually link in any _new_ intermediate
+	// nodes. We'll do that on return if nothing goes wrong.
+	ln := n.links[i/nfh]
+	if ln == nil {
+		ln = &link{cached: new(node)}
+	}
+	subn, err := ln.load(ctx, bs, height-1)
+	if err != nil {
+		return false, err
+	}
+
+	nodeAdded, err := subn.set(ctx, bs, height-1, i%nfh, val)
+	if err != nil {
+		return false, err
+	}
+
+	// Make all modifications on the way back up if there was no error.
+	ln.dirty = true // only mark dirty on success.
+	n.links[i/nfh] = ln
+
+	return nodeAdded, nil
+}
+
+func (n *node) flush(ctx context.Context, bs cbor.IpldStore, height int) (*internal.Node, error) {
+	var nd internal.Node
+	if height == 0 {
+		for i, val := range n.values {
+			if val == nil {
+				continue
+			}
+			nd.Values = append(nd.Values, val)
+			nd.Bmap[i/8] |= 1 << (uint(i) % 8)
+		}
+		return &nd, nil
+	}
+
+	for i, ln := range n.links {
+		if ln == nil {
+			continue
+		}
+		if ln.dirty {
+			subn, err := ln.cached.flush(ctx, bs, height-1)
```

Comment:
I don't think there is any possible code path that can hit this, but if the dirty bit is somehow set without a cached node then this will panic.  What are your thoughts on checking for this condition and returning an error in this case?  Is panicing preferable?

Reply:
It should be an invariant and I'd usually panic in cases like this. But given the blockchain context, it probably makes sense to be extra defensive.